### PR TITLE
Introduce firewall-controller-manager.

### PR DESCRIPTION
**Pull request by Gerrit91**
This introduces the firewall-controller-manager as part of https://github.com/metal-stack/gardener-extension-provider-metal/issues/276.

This makes all existing firewalls getting adopted by the new [firewall-controller-manager](https://github.com/metal-stack/firewall-controller-manager). The old `Firewall/v1` resource will still be deployed such that old firewall-controller versions can still continue to run. 

# Required Actions

```ACTIONS_REQUIRED
Clusters provisioned with firewall-images that do not contain a firewall-controller v2 require an additional shoot reconciliation to succeed, increasing the provisioning time.
```

### CODE REVIEW

File: pkg/controller/infrastructure/actuator_reconcile.go [hunk 1]
```diff
@@ -109,299 +63,43 @@ func (a *actuator) Reconcile(ctx context.Context, infrastructure *extensionsv1al
...
-			RequeueAfter: 30 * time.Second,
-		}
 	}
-
-	err = updateProviderStatus(ctx, r.c, r.infrastructure, r.providerStatus, &nodeCIDR)
+	nodeCIDR, err := ensureNodeNetwork(ctx, networkReconciler)
 	if err != nil {
-		r.logger.Error(err, "firewalls update provider status", "nodecidr", nodeCIDR)
-		return "", "", &reconciler.RequeueAfterError{
+		return &reconciler.RequeueAfterError{
```
Comment (vknabel) [inline pkg/controller/infrastructure/actuator_reconcile.go:76]: In other cases we removed the `*reconciler.RequeueAfterError` as it was only printed but not executed. Does it work in this case? (and / or in other occurrences)

File: pkg/controller/worker/machine_dependencies.go [hunk 2]
```diff
@@ -1,13 +1,72 @@
...
 func (w *workerDelegate) DeployMachineDependencies(_ context.Context) error {
 	return nil
 }
 
 // CleanupMachineDependencies implements genericactuator.WorkerDelegate.
-func (w *workerDelegate) CleanupMachineDependencies(_ context.Context) error {
-	return nil
+func (w *workerDelegate) CleanupMachineDependencies(ctx context.Context) error {
+	worker := w.worker.DeepCopy()
+	err := w.client.Get(ctx, client.ObjectKeyFromObject(w.worker), worker)
```
Comment (vknabel) [inline pkg/controller/worker/machine_dependencies.go:25]: Is this still necessary?

File: pkg/controller/worker/machines.go [hunk 3]
```diff
@@ -217,3 +264,340 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
 
 	return nil
 }
+
+// migrateFirewall can be removed along with the deployment of the old firewall resource after all firewall are running firewall-controller >= v0.2.0
```
Comment (vknabel) [inline pkg/controller/worker/machines.go:268]:
```suggestion
// migrateFirewall can be removed along with the deployment of the old firewall resource after all firewalls are running firewall-controller >= v0.2.0
```

File: pkg/controller/worker/machines.go [hunk 4]
```diff
@@ -217,3 +264,340 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
...
+				Namespace: namespace,
+			},
+		}
+
+		_, err = controllerutil.CreateOrUpdate(ctx, w.client, f, func() error {
+			f.Labels = map[string]string{
+				tag.ClusterID: clusterID,
+			}
+			if v, err := semver.NewVersion(fwcv.Version); err == nil && v.LessThan(semver.MustParse("v2.0.0")) {
+				f.Annotations = map[string]string{
```
Comment (vknabel) [inline pkg/controller/worker/machines.go:359]: Here we discard all annotations. Should we add `fcmv2.FirewallNoControllerConnectionAnnotation` instead? (check for nil)

File: pkg/controller/infrastructure/actuator_reconcile.go (hunk 1 above)
Comment (Gerrit91) [reply pkg/controller/infrastructure/actuator_reconcile.go:76]: Seems like I kept the `RequeueAfterError` in the `Reconcile` funcs and just returned errors from the functions are called in `Reconcile`. Despite I have never seen a requeue really taking place, when looking at the underlying code of the extension library, the requeue error should work. So I would assume implementation is ok.

File: pkg/controller/worker/machine_dependencies.go (hunk 2 above)
Comment (Gerrit91) [reply pkg/controller/worker/machine_dependencies.go:25]: I will try it again without re-fetching the worker resource.

**Review by majst01 — APPROVED** [review]

**Pull request merged**